### ISSUE TITLE
NullPointerException in HttpFeatureFlagFetcher

### ISSUE THREAD

**Issue by wokkaflokka** [launchdarkly/android-client#28]
### Notes 

In production, we are seeing the following crash at the rate of 1 per hour.

```
Caused by java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String com.launchdarkly.android.LDUser.getAsUrlSafeBase64()' on a null object reference
       at com.launchdarkly.android.HttpFeatureFlagFetcher.fetch(HttpFeatureFlagFetcher.java:69)
       at com.launchdarkly.android.UserManager.updateCurrentUser(UserManager.java:133)
       at com.launchdarkly.android.PollingUpdater.onReceive(PollingUpdater.java:34)
       at android.app.ActivityThread.handleReceiver(ActivityThread.java:2838)
       at android.app.ActivityThread.access$1700(ActivityThread.java:178)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1547)
       at android.os.Handler.dispatchMessage(Handler.java:111)
       at android.os.Looper.loop(Looper.java:194)
       at android.app.ActivityThread.main(ActivityThread.java:5637)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:959)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:754)
```

From the stacktrace, the apparent codepath through LD is:

[UserManager.java:133](https://github.com/launchdarkly/android-client/blob/master/launchdarkly-android-client/src/main/java/com/launchdarkly/android/UserManager.java#L133)
[HttpFeatureFlagFetcher.java:69](https://github.com/launchdarkly/android-client/blob/master/launchdarkly-android-client/src/main/java/com/launchdarkly/android/HttpFeatureFlagFetcher.java#L69)

### Steps to Reproduce

Unknown at this time.

### Expected Behavior

0 crashes from LaunchDarkly SDK

**Comment by arun251** [comment]
Hi @wokkaflokka ,

Thanks for reporting the issue. Is it possible that a null value is getting passed as the `user` argument for `LDClient.init`?

**Comment by wokkaflokka** [comment]
@arun251 

I don't have data that allows me to conclude it's impossible a null value is ever provided, but based on our configuration it seems unlikely. 

We create the user as follows: <img width="715" alt="screen shot 2017-09-28 at 10 52 48" src="https://user-images.githubusercontent.com/995007/30977124-3eea2b50-a43c-11e7-8020-ec6ea02c3e8a.png">

We register the user with LaunchDarkly as follows: <img width="346" alt="screen shot 2017-09-28 at 11 01 11" src="https://user-images.githubusercontent.com/995007/30977349-e8f434e2-a43c-11e7-971d-75c369622959.png">

For user's that sign-out in the context of a given session, we set an anonymous user as "active":
 <img width="389" alt="screen shot 2017-09-28 at 10 54 11" src="https://user-images.githubusercontent.com/995007/30977128-42b93672-a43c-11e7-8530-a8b74d81fc77.png">

A general question I have in response to yours would be: if it is the case that a null user is provided as an argument, does it make sense to fail in this way? From what I can see, this crash is manifesting from a `BroadcastReceiver` -- indicating that the failure is happening asynchronously from any potential `init` or `identify` call. If a null user is invalid input, perhaps it would make sense to raise an exception synchronously from the calling method in that event.

**Comment by arun251** [comment]
I absolutely agree that we should be failing fast for invalid input. We will make that change.

Just so we can confirm that such a change will resolve the issue, can you check whether the call to `LDClient.init` could ever include a null value for `LDUser`?

**Comment by wokkaflokka** [comment]
Based on the code posted above, I don't think it's possible that we are providing a null value. 

Is there a way to prioritize getting this issue addressed given that we have an enterprise contract for our account? We have this deployed to a non-trivial audience for Android alone (>100K), and the nature of these crashes is driving us to evaluate using the provided android SDK versus hitting the API directly.

**Comment by arun251** [comment]
Hi @wokkaflokka,
I apologize for the delay, but we've just released version 2.1.0, which includes early validation of arguments passed to `init`. Please give that a spin. If you continue to experience an issue, we can work together more closely if you contact support@launchdarkly.com.

**Comment by wokkaflokka** [comment]
@arun251 thanks, will take a look. I have a meeting with your team on Friday, I will evaluate this change before hand. A cursory inspection of #29 suggests it may indeed be fixed.

**Comment by wokkaflokka** [comment]
@arun251 

I have tested this update in production and it indeed stops the crash we were seeing.

I'm not seeing any error's related to calling `LDClient.init` with bad input. Noting our code below, if we were invoking this with bad input, per the SDK code it should raise an exception which, in our error subscriber's implementation, would log the error to Fabric. I think this supports my findings above -- that there is a condition where `HttpFeatureFlagFetcher#fetch` is invoked with null input. 

<img width="735" alt="screen shot 2017-10-20 at 10 14 23" src="https://user-images.githubusercontent.com/995007/31828459-eb846b60-b57f-11e7-94c7-9438065a94f3.png">

**Comment by arun251** [comment]
It sounds like this issue has been resolved for you in 2.1.0. Let me know if you see a recurrence.
